client_servers.go

```diff
@@ -0,0 +1,686 @@
+	req := c.newRequest("POST", fmt.Sprintf("/servers/%s/databases/%s/rotate-password", identifier, id), nil)
+	res, err := c.Http.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err := validate(res)
+	if err != nil {
+		return nil, err
+	}
+
+	var model struct {
+		Attributes ClientDatabase `json:"attributes"`
+	}
+	if err = json.Unmarshal(buf, &model); err != nil {
+		return nil, err
+	}
+
+	return &model.Attributes, nil
+}
+
+func (c *Client) DeleteDatabase(identifier, id string) error {
+	req := c.newRequest("DELETE", fmt.Sprintf("/servers/%s/databases/%s", identifier, id), nil)
+	res, err := c.Http.Do(req)
+	if err != nil {
+		return err
+	}
+
+	_, err = validate(res)
+	return err
+}
+
+type File struct {
+	Name       string     `json:"name"`
+	Mode       string     `json:"mode"`
+	ModeBits   uint       `json:"mode_bits"`
+	Size       int64      `json:"size"`
+	IsFile     bool       `json:"is_file"`
+	IsSymlink  bool       `json:"is_symlink"`
+	MimeType   string     `json:"mime_type"`
+	CreatedAt  *time.Time `json:"created_at"`
+	ModifiedAt *time.Time `json:"modified_at,omitempty"`
+}
+
+func (c *Client) ServerFiles(identififer, root string) ([]*File, error) {
+	req := c.newRequest("GET", fmt.Sprintf("/servers/%s/files?directory=%s", identififer, root), nil)
+	res, err := c.Http.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err := validate(res)
+	if err != nil {
+		return nil, err
+	}
+
+	var model struct {
+		Data []struct {
+			Attributes *File `json:"attributes"`
+		} `json:"data"`
+	}
+	if err = json.Unmarshal(buf, &model); err != nil {
+		return nil, err
+	}
+
+	files := make([]*File, 0, len(model.Data))
+	for _, f := range model.Data {
+		files = append(files, f.Attributes)
+	}
+
+	return files, nil
+}
+
+func (c *Client) ServerFileContents(identifier, file string) ([]byte, error) {
+	req := c.newRequest("GET", fmt.Sprintf("/servers/%s/files/contents?file=%s", identifier, url.PathEscape(file)), nil)
+	req.Header.Set("Accept", "application/json,text/plain")
+
+	res, err := c.Http.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	return validate(res)
+}
+
+type Downloader struct {
+	client *Client
+	Name   string
+	Path   string
+	url    string
+}
+
+func (d *Downloader) Client() *Client {
+	return d.client
+}
+
+func (d *Downloader) URL() string {
+	return d.url
+}
+
+func (d *Downloader) Execute() error {
+	info, err := os.Stat(d.Path)
+	if err == nil {
+		if !info.IsDir() {
+			return errors.New("refusing to overwrite existing file path")
+		}
+	}
+
+	file, err := os.OpenFile(d.Name, os.O_CREATE|os.O_RDWR|os.O_TRUNC, 0o644)
+	if err != nil {
+		return err
+	}
+
+	res, err := d.client.Http.Get(d.URL())
+	if err != nil {
+		return err
+	}
+
+	if res.StatusCode != 200 {
+		return fmt.Errorf("recieved an unexpected response: %s", res.Status)
+	}
+
+	defer res.Body.Close()
+	io.Copy(file, res.Body)
+	return nil
+}
+
+func (c *Client) DownloadServerFile(identifier, file string) (*Downloader, error) {
```

Comment:
This fails if the `mimetype` is `inode/directory`. 

You either want to throw an error, or run a compress and then download the compressed file.

Reply:
I've fixed it creating the empty file on error, but I couldn't find an effective way to check the MIME type so it will still return the 404 error it gets from Wings.

Reply:
you might be better off passing the entire file off to this function then. It will have all the context then.

Reply:
I've set it to check the file first before continuing with the rest of the function.